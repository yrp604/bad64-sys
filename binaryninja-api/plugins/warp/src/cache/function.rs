use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
