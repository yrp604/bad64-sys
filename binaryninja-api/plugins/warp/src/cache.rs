use crate::convert::{from_bn_symbol, from_bn_type_internal};
use crate::{build_function, function_guid};
use binaryninja::architecture::Architecture;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::function::Function as BNFunction;
use binaryninja::low_level_il::function::{
    FunctionMutability, LowLevelILFunction, NonSSA, RegularNonSSA,
};
use binaryninja::low_level_il::RegularLowLevelILFunction;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::symbol::Symbol as BNSymbol;
use binaryninja::types::NamedTypeReference as BNNamedTypeReference;
use binaryninja::ObjectDestructor;
use dashmap::mapref::one::Ref;
use dashmap::DashMap;
use std::collections::HashSet;
use std::hash::{DefaultHasher, Hash, Hasher};
use std::sync::OnceLock;
use warp::r#type::ComputedType;
use warp::signature::function::constraints::FunctionConstraint;
use warp::signature::function::{Function, FunctionGUID};

pub static MATCHED_FUNCTION_CACHE: OnceLock<DashMap<ViewID, MatchedFunctionCache>> =
    OnceLock::new();
pub static FUNCTION_CACHE: OnceLock<DashMap<ViewID, FunctionCache>> = OnceLock::new();
pub static GUID_CACHE: OnceLock<DashMap<ViewID, GUIDCache>> = OnceLock::new();
pub static TYPE_REF_CACHE: OnceLock<DashMap<ViewID, TypeRefCache>> = OnceLock::new();

pub fn register_cache_destructor() {
    pub static mut CACHE_DESTRUCTOR: CacheDestructor = CacheDestructor;
    #[allow(static_mut_refs)]
    // SAFETY: This can be done as the backing data is an opaque ZST.
    unsafe {
        CACHE_DESTRUCTOR.register()
    };
}

pub fn cached_function_match<F>(function: &BNFunction, f: F) -> Option<Function>
where
    F: Fn() -> Option<Function>,
{
    let view = function.view();
    let view_id = ViewID::from(view.as_ref());
    let function_id = FunctionID::from(function);
    let function_cache = MATCHED_FUNCTION_CACHE.get_or_init(Default::default);
    match function_cache.get(&view_id) {
        Some(cache) => cache.get_or_insert(&function_id, f).to_owned(),
        None => {
            let cache = MatchedFunctionCache::default();
            let matched = cache.get_or_insert(&function_id, f).to_owned();
            function_cache.insert(view_id, cache);
            matched
        }
    }
}

pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let function_id = FunctionID::from(function);
    let function_cache = MATCHED_FUNCTION_CACHE.get_or_init(Default::default);
    function_cache
        .get(&view_id)?
        .get(&function_id)?
        .value()
        .to_owned()
}

pub fn cached_function<A: Architecture>(
    function: &BNFunction,
    llil: &RegularLowLevelILFunction<A>,
) -> Function {
    let view = function.view();
    let view_id = ViewID::from(view.as_ref());
    let function_cache = FUNCTION_CACHE.get_or_init(Default::default);
    match function_cache.get(&view_id) {
        Some(cache) => cache.function(function, llil),
        None => {
            let cache = FunctionCache::default();
            let function = cache.function(function, llil);
            function_cache.insert(view_id, cache);
            function
        }
    }
}

pub fn cached_call_site_constraints(function: &BNFunction) -> HashSet<FunctionConstraint> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.call_site_constraints(function),
        None => {
            let cache = GUIDCache::default();
            let constraints = cache.call_site_constraints(function);
            guid_cache.insert(view_id, cache);
            constraints
        }
    }
}

pub fn cached_adjacency_constraints<F>(
    function: &BNFunction,
    filter: F,
) -> HashSet<FunctionConstraint>
where
    F: Fn(&BNFunction) -> bool,
{
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.adjacency_constraints(function, filter),
        None => {
            let cache = GUIDCache::default();
            let constraints = cache.adjacency_constraints(function, filter);
            guid_cache.insert(view_id, cache);
            constraints
        }
    }
}

pub fn cached_function_guid<A: Architecture, M: FunctionMutability>(
    function: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> FunctionGUID {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.function_guid(function, llil),
        None => {
            let cache = GUIDCache::default();
            let guid = cache.function_guid(function, llil);
            guid_cache.insert(view_id, cache);
            guid
        }
    }
}

pub fn try_cached_function_guid(function: &BNFunction) -> Option<FunctionGUID> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    guid_cache.get(&view_id)?.try_function_guid(function)
}

pub fn cached_type_reference(
    view: &BinaryView,
    visited_refs: &mut HashSet<TypeRefID>,
    type_ref: &BNNamedTypeReference,
) -> Option<ComputedType> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    match type_ref_cache.get(&view_id) {
        Some(cache) => cache.cached_type_reference(view, visited_refs, type_ref),
        None => {
            let cache = TypeRefCache::default();
            let ntr = cache.cached_type_reference(view, visited_refs, type_ref);
            type_ref_cache.insert(view_id, cache);
            ntr
        }
    }
}

pub fn cached_type_references(view: &BinaryView) -> Option<Ref<ViewID, TypeRefCache>> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    type_ref_cache.get(&view_id)
}

#[derive(Clone, Debug, Default)]
pub struct MatchedFunctionCache {
    pub cache: DashMap<FunctionID, Option<Function>>,
}

impl MatchedFunctionCache {
    pub fn get_or_insert<F>(
        &self,
        function_id: &FunctionID,
        f: F,
    ) -> Ref<'_, FunctionID, Option<Function>>
    where
        F: FnOnce() -> Option<Function>,
    {
        self.cache.get(function_id).unwrap_or_else(|| {
            self.cache.insert(*function_id, f());
            self.cache.get(function_id).unwrap()
        })
    }

    pub fn get(&self, function_id: &FunctionID) -> Option<Ref<'_, FunctionID, Option<Function>>> {
        self.cache.get(function_id)
    }
}

#[derive(Clone, Debug, Default)]
pub struct FunctionCache {
    pub cache: DashMap<FunctionID, Function>,
}

impl FunctionCache {
    pub fn function<A: Architecture>(
        &self,
        function: &BNFunction,
        llil: &RegularLowLevelILFunction<A>,
    ) -> Function {
        let function_id = FunctionID::from(function);
        match self.cache.get(&function_id) {
            Some(function) => function.value().to_owned(),
            None => {
                let function = build_function(function, llil);
                self.cache.insert(function_id, function.clone());
                function
            }
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct GUIDCache {
    pub cache: DashMap<FunctionID, FunctionGUID>,
}

impl GUIDCache {
    pub fn call_site_constraints(&self, function: &BNFunction) -> HashSet<FunctionConstraint> {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let func_platform = function.platform();
        let mut constraints = HashSet::new();
        for call_site in &function.call_sites() {
            for cs_ref_addr in view.code_refs_from_addr(call_site.address, Some(function)) {
                match view.function_at(&func_platform, cs_ref_addr) {
                    Some(cs_ref_func) => {
                        // Call site is a function, constrain on it.
                        let cs_ref_func_id = FunctionID::from(cs_ref_func.as_ref());
                        if cs_ref_func_id != func_id {
                            let call_site_offset: i64 =
                                call_site.address.wrapping_sub(func_start) as i64;
                            // TODO: If the function is thunk we should also insert the called function.
                            constraints
                                .insert(self.function_constraint(&cs_ref_func, call_site_offset));
                        }
                    }
                    None => {
                        // We could be dealing with an extern symbol, get the symbol as a constraint.
                        let call_site_offset: i64 =
                            call_site.address.wrapping_sub(func_start) as i64;
                        if let Some(call_site_sym) = view.symbol_by_address(cs_ref_addr) {
                            constraints.insert(
                                self.function_constraint_from_symbol(
                                    &call_site_sym,
                                    call_site_offset,
                                ),
                            );
                        }
                    }
                }
            }
        }
        constraints
    }

    pub fn adjacency_constraints<F>(
        &self,
        function: &BNFunction,
        filter: F,
    ) -> HashSet<FunctionConstraint>
    where
        F: Fn(&BNFunction) -> bool,
    {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let mut constraints = HashSet::new();

        let mut func_addr_constraint = |func_start_addr| {
            // NOTE: We could potentially have dozens of functions all at the same start address.
            for curr_func in &view.functions_at(func_start_addr) {
                let curr_func_id = FunctionID::from(curr_func.as_ref());
                if curr_func_id != func_id && filter(curr_func.as_ref()) {
                    // NOTE: For this to work the GUID has to have already been cached. If not it will just be the symbol.
                    // Function adjacent to another function, constrain on the pattern.
                    let curr_addr_offset = (func_start_addr as i64) - func_start as i64;
                    constraints.insert(self.function_constraint(&curr_func, curr_addr_offset));
                }
            }
        };

        let mut before_func_start = func_start;
        for _ in 0..2 {
            before_func_start = view.function_start_before(before_func_start);
            func_addr_constraint(before_func_start);
        }

        let mut after_func_start = func_start;
        for _ in 0..2 {
            after_func_start = view.function_start_after(after_func_start);
            func_addr_constraint(after_func_start);
        }

        constraints
    }

    /// Construct a function constraint, must pass the offset at which it is located.
    pub fn function_constraint(&self, function: &BNFunction, offset: i64) -> FunctionConstraint {
        let guid = self.try_function_guid(function);
        let symbol = from_bn_symbol(&function.symbol());
        FunctionConstraint {
            guid,
            symbol: Some(symbol),
            offset,
        }
    }

    /// Construct a function constraint from a symbol, typically used for extern function call sites, must pass the offset at which it is located.
    pub fn function_constraint_from_symbol(
        &self,
        symbol: &BNSymbol,
        offset: i64,
    ) -> FunctionConstraint {
        let symbol = from_bn_symbol(symbol);
        FunctionConstraint {
            guid: None,
            symbol: Some(symbol),
            offset,
        }
    }

    pub fn function_guid<A: Architecture, M: FunctionMutability>(
        &self,
        function: &BNFunction,
        llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
    ) -> FunctionGUID {
        let function_id = FunctionID::from(function);
        match self.cache.get(&function_id) {
            Some(function_guid) => function_guid.value().to_owned(),
            None => {
                let function_guid = function_guid(function, llil);
                self.cache.insert(function_id, function_guid);
                function_guid
            }
        }
    }

    pub fn try_function_guid(&self, function: &BNFunction) -> Option<FunctionGUID> {
        let function_id = FunctionID::from(function);
        self.cache
            .get(&function_id)
            .map(|function_guid| function_guid.value().to_owned())
    }
}

#[derive(Clone, Debug, Default)]
pub struct TypeRefCache {
    pub cache: DashMap<TypeRefID, Option<ComputedType>>,
}

impl TypeRefCache {
    /// NOTE: No self-referential type must be used on this function.
    pub fn cached_type_reference(
        &self,
        view: &BinaryView,
        visited_refs: &mut HashSet<TypeRefID>,
        type_ref: &BNNamedTypeReference,
    ) -> Option<ComputedType> {
        let ntr_id = TypeRefID::from(type_ref);
        match self.cache.get(&ntr_id) {
            Some(cache) => cache.to_owned(),
            None => match type_ref.target(view) {
                Some(raw_ty) => {
                    let computed_ty = ComputedType::new(from_bn_type_internal(
                        view,
                        visited_refs,
                        &raw_ty,
                        MAX_CONFIDENCE,
                    ));
                    self.cache
                        .entry(ntr_id)
                        .insert(Some(computed_ty))
                        .to_owned()
                }
                None => self.cache.entry(ntr_id).insert(None).to_owned(),
            },
        }
    }
}

/// A unique view ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct ViewID(u64);

impl From<&BinaryView> for ViewID {
    fn from(value: &BinaryView) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.original_image_base());
        hasher.write_usize(value.file().session_id());
        Self(hasher.finish())
    }
}

impl From<BNRef<BinaryView>> for ViewID {
    fn from(value: BNRef<BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BinaryView>> for ViewID {
    fn from(value: Guard<'_, BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique function ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct FunctionID(u64);

impl From<&BNFunction> for FunctionID {
    fn from(value: &BNFunction) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.start());
        hasher.write_u64(value.lowest_address());
        hasher.write_u64(value.highest_address());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNFunction>> for FunctionID {
    fn from(value: BNRef<BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNFunction>> for FunctionID {
    fn from(value: Guard<'_, BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique named type reference ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct TypeRefID(u64);

impl From<&BNNamedTypeReference> for TypeRefID {
    fn from(value: &BNNamedTypeReference) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write(value.id().as_bytes());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNNamedTypeReference>> for TypeRefID {
    fn from(value: BNRef<BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNNamedTypeReference>> for TypeRefID {
    fn from(value: Guard<'_, BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}

pub struct CacheDestructor;

impl ObjectDestructor for CacheDestructor {
    fn destruct_view(&self, view: &BinaryView) {
        // Clear caches as the view is no longer alive.
        let view_id = ViewID::from(view);
        if let Some(cache) = MATCHED_FUNCTION_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = FUNCTION_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = GUID_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = TYPE_REF_CACHE.get() {
            cache.remove(&view_id);
        }
        log::debug!("Removed WARP caches for {:?}", view.file().filename());
    }
}
