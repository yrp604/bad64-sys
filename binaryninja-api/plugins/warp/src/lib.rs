use crate::cache::{
    cached_adjacency_constraints, cached_call_site_constraints, cached_function_guid,
};
use crate::convert::{from_bn_symbol, from_bn_type};
use binaryninja::architecture::{
    Architecture, ImplicitRegisterExtend, Register as BNRegister, RegisterInfo,
};
use binaryninja::basic_block::BasicBlock as BNBasicBlock;
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::function::{Function as BNFunction, NativeBlock};
use binaryninja::low_level_il::expression::{ExpressionHandler, LowLevelILExpressionKind};
use binaryninja::low_level_il::function::{
    FunctionMutability, LowLevelILFunction, NonSSA, RegularNonSSA,
};
use binaryninja::low_level_il::instruction::{
    InstructionHandler, LowLevelILInstruction, LowLevelILInstructionKind,
};
use binaryninja::low_level_il::{LowLevelILRegister, VisitorAction};
use binaryninja::rc::Ref as BNRef;
use std::path::PathBuf;
use warp::signature::basic_block::BasicBlockGUID;
use warp::signature::function::constraints::FunctionConstraints;
use warp::signature::function::{Function, FunctionGUID};

pub mod cache;
pub mod convert;
mod matcher;
/// Only used when compiled for cdylib target.
mod plugin;

pub fn core_signature_dir() -> PathBuf {
    // Get core signatures for the given platform
    let install_dir = binaryninja::install_directory();
    // macOS core dir is separate from the install dir.
    #[cfg(target_os = "macos")]
    let core_dir = install_dir.parent().unwrap().join("Resources");
    #[cfg(not(target_os = "macos"))]
    let core_dir = install_dir;
    core_dir.join("signatures")
}

pub fn user_signature_dir() -> PathBuf {
    binaryninja::user_directory().join("signatures/")
}

pub fn build_function<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> Function {
    let bn_fn_ty = func.function_type();
    Function {
        guid: cached_function_guid(func, llil),
        symbol: from_bn_symbol(&func.symbol()),
        ty: from_bn_type(&func.view(), &bn_fn_ty, MAX_CONFIDENCE),
        constraints: FunctionConstraints {
            // NOTE: Adding adjacent only works if analysis is complete.
            // NOTE: We do not filter out adjacent functions here.
            adjacent: cached_adjacency_constraints(func, |_| true),
            call_sites: cached_call_site_constraints(func),
            // TODO: Add caller sites (when adjacent and call sites are minimal)
            // NOTE: Adding caller sites only works if analysis is complete.
            caller_sites: Default::default(),
        },
    }
}

/// Basic blocks sorted from high to low.
pub fn sorted_basic_blocks(func: &BNFunction) -> Vec<BNRef<BNBasicBlock<NativeBlock>>> {
    let mut basic_blocks = func
        .basic_blocks()
        .iter()
        .map(|bb| bb.clone())
        .collect::<Vec<_>>();
    basic_blocks.sort_by_key(|f| f.start_index());
    basic_blocks
}

pub fn function_guid<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> FunctionGUID {
    let basic_blocks = sorted_basic_blocks(func);
    let basic_block_guids = basic_blocks
        .iter()
        .map(|bb| basic_block_guid(bb, llil))
        .collect::<Vec<_>>();
    FunctionGUID::from_basic_blocks(&basic_block_guids)
}

pub fn basic_block_guid<A: Architecture, M: FunctionMutability>(
    basic_block: &BNBasicBlock<NativeBlock>,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> BasicBlockGUID {
    let func = basic_block.function();
    let view = func.view();
    let arch = func.arch();
    let max_instr_len = arch.max_instr_len();

    // NOPs and useless moves are blacklisted to allow for hot-patchable functions.
    let is_blacklisted_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        match instr.kind() {
            LowLevelILInstructionKind::Nop(_) => true,
            LowLevelILInstructionKind::SetReg(op) => {
                match op.source_expr().kind() {
                    LowLevelILExpressionKind::Reg(source_op)
                        if op.dest_reg() == source_op.source_reg() =>
                    {
                        match op.dest_reg() {
                            LowLevelILRegister::ArchReg(r) => {
                                // If this register has no implicit extend then we can safely assume it's a NOP.
                                // Ex. on x86_64 we don't want to remove `mov edi, edi` as it will zero the upper 32 bits.
                                // Ex. on x86 we do want to remove `mov edi, edi` as it will not have a side effect like above.
                                matches!(
                                    r.info().implicit_extend(),
                                    ImplicitRegisterExtend::NoExtend
                                )
                            }
                            LowLevelILRegister::Temp(_) => false,
                        }
                    }
                    _ => false,
                }
            }
            _ => false,
        }
    };

    let is_variant_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        let is_variant_expr = |expr: &LowLevelILExpressionKind<A, M, NonSSA<RegularNonSSA>>| {
            // TODO: Checking the section here is slow, we should gather all section ranges outside of this.
            match expr {
                LowLevelILExpressionKind::ConstPtr(op)
                    if !view.sections_at(op.value()).is_empty() =>
                {
                    // Constant Pointer must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                LowLevelILExpressionKind::ExternPtr(_) => true,
                LowLevelILExpressionKind::Const(op) if !view.sections_at(op.value()).is_empty() => {
                    // Constant value must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                _ => false,
            }
        };

        // Visit instruction expressions looking for variant expression, [VisitorAction::Halt] means variant.
        instr.visit_tree(&mut |expr| {
            if is_variant_expr(&expr.kind()) {
                // Found a variant expression
                VisitorAction::Halt
            } else {
                VisitorAction::Descend
            }
        }) == VisitorAction::Halt
    };

    let basic_block_range = basic_block.start_index()..basic_block.end_index();
    let mut basic_block_bytes = Vec::with_capacity(basic_block_range.count());
    for instr_addr in basic_block.into_iter() {
        let mut instr_bytes = view.read_vec(instr_addr, max_instr_len);
        if let Some(instr_info) = arch.instruction_info(&instr_bytes, instr_addr) {
            instr_bytes.truncate(instr_info.length);
            if let Some(instr_llil) = llil.instruction_at(instr_addr) {
                // If instruction is blacklisted don't include the bytes.
                if !is_blacklisted_instr(&instr_llil) {
                    if is_variant_instr(&instr_llil) {
                        // Found a variant instruction, mask off entire instruction.
                        instr_bytes.fill(0);
                    }
                    // Add the instructions bytes to the basic blocks bytes
                    basic_block_bytes.extend(instr_bytes);
                }
            }
        }
    }

    BasicBlockGUID::from(basic_block_bytes.as_slice())
}

#[cfg(test)]
mod tests {
    use crate::cache::cached_function_guid;
    use binaryninja::binary_view::BinaryViewExt;
    use binaryninja::headless::Session;
    use std::path::PathBuf;
    use std::sync::OnceLock;

    static INIT: OnceLock<Session> = OnceLock::new();

    fn get_session<'a>() -> &'a Session {
        // TODO: This is not shared between other test modules, should still be fine (mutex in core now).
        INIT.get_or_init(|| Session::new().expect("Failed to initialize session"))
    }

    #[test]
    fn insta_signatures() {
        let session = get_session();
        let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
        for entry in std::fs::read_dir(out_dir).expect("Failed to read OUT_DIR") {
            let entry = entry.expect("Failed to read directory entry");
            let path = entry.path();
            if path.is_file() {
                let view = session.load(&path).expect("Failed to load view");
                let mut functions = view
                    .functions()
                    .iter()
                    .map(|f| cached_function_guid(&f, &f.low_level_il().unwrap()))
                    .collect::<Vec<_>>();
                functions.sort_by_key(|guid| guid.guid);
                let snapshot_name =
                    format!("snapshot_{}", path.file_stem().unwrap().to_string_lossy());
                insta::assert_debug_snapshot!(snapshot_name, functions);
            }
        }
    }
}
