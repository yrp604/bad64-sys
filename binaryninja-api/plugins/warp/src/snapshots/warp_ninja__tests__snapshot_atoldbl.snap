---
source: plugins/warp/src/lib.rs
expression: functions
---
[
    FunctionGUID {
        guid: 02d864ef-80bc-5265-b78f-0e42815a4d1d,
    },
    FunctionGUID {
        guid: 07b3f23f-333b-5c13-873f-5f1973cee8e2,
    },
    FunctionGUID {
        guid: 13b16f81-0c6f-5aee-ad9f-0d142658cf18,
    },
    FunctionGUID {
        guid: 1de88f85-a19c-5454-9122-ddc80f7509b4,
    },
    FunctionGUID {
        guid: 1e794537-6289-59e7-bef9-0c72f3989db8,
    },
    FunctionGUID {
        guid: 22c32840-7bb0-5d80-858c-55dd04553db0,
    },
    FunctionGUID {
        guid: 29690354-fa27-54d1-a8be-18535b19b1c3,
    },
    FunctionGUID {
        guid: 29690354-fa27-54d1-a8be-18535b19b1c3,
    },
    FunctionGUID {
        guid: 30f44555-ef86-5e85-96c3-6153f4bcc11e,
    },
    FunctionGUID {
        guid: 364c5864-0cd1-5527-8ca0-2378d5b4c0a5,
    },
    FunctionGUID {
        guid: 39cd0d51-d688-58e2-a757-ac868ff8635e,
    },
    FunctionGUID {
        guid: 3c83a5f5-215d-5d25-97c8-ffe65f73907c,
    },
    FunctionGUID {
        guid: 3cc2e827-b707-5c9d-82e9-d76ac3abc904,
    },
    FunctionGUID {
        guid: 41fd6d24-0f8b-5848-9873-65b98bd370ab,
    },
    FunctionGUID {
        guid: 43a8c54b-dd4f-5334-ae00-218cb758ad4c,
    },
    FunctionGUID {
        guid: 45770948-1c56-5b24-bd38-f5a559ae8d22,
    },
    FunctionGUID {
        guid: 4738dcb9-abab-5157-baef-8edf38d9fabc,
    },
    FunctionGUID {
        guid: 48551d5f-2e1f-557b-85a3-a122d3ce07cf,
    },
    FunctionGUID {
        guid: 4ba8a311-60f3-57d4-a17a-c0cc93f5c2a4,
    },
    FunctionGUID {
        guid: 4f645f3c-3e89-5b5d-a185-9f1a25e94ed2,
    },
    FunctionGUID {
        guid: 5779439c-a22f-5d33-8868-4e1832ad49e9,
    },
    FunctionGUID {
        guid: 5b7f41e9-de1e-558c-ac7d-888acca3a76c,
    },
    FunctionGUID {
        guid: 61d32a04-c077-5985-b241-c326dd5b1864,
    },
    FunctionGUID {
        guid: 6444ff14-38c5-520e-b9e3-6e3ae15215c4,
    },
    FunctionGUID {
        guid: 6589669f-6b11-569e-9adf-ca3c6d6bc62e,
    },
    FunctionGUID {
        guid: 6590d978-d2b0-510c-b5a6-a81361800609,
    },
    FunctionGUID {
        guid: 67f303c6-d2e4-55e1-8a5f-5492da69e69f,
    },
    FunctionGUID {
        guid: 6abe3fc2-6d29-5fde-a31e-bb8249db6bf9,
    },
    FunctionGUID {
        guid: 6ef93fa8-909e-596c-9454-74b89e8cabb7,
    },
    FunctionGUID {
        guid: 7995ca73-04fe-5b30-9e88-1e69197007c1,
    },
    FunctionGUID {
        guid: 7c43c697-22ef-57a9-a43a-2214e627749d,
    },
    FunctionGUID {
        guid: 86ab6ab4-45e5-5c08-9a47-05344ea503ab,
    },
    FunctionGUID {
        guid: 88d9b26b-3884-58d6-b164-435d8d888e5c,
    },
    FunctionGUID {
        guid: 8a2b7bda-5fdb-5ba3-888b-20d5b8d7b2cd,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8c736b66-b63f-5000-a060-2e60a1b8952a,
    },
    FunctionGUID {
        guid: 8ccf8b37-65c5-525a-b96a-64f26d711c75,
    },
    FunctionGUID {
        guid: 9676fc90-ecba-595b-8d24-5df4e66ac683,
    },
    FunctionGUID {
        guid: 9891b84c-000b-529a-8c5b-d762120eff9d,
    },
    FunctionGUID {
        guid: a030f6f1-e29b-5e4a-8f83-56e6c6c51c15,
    },
    FunctionGUID {
        guid: a23505e4-e7cd-5543-897c-46b97c1eaef3,
    },
    FunctionGUID {
        guid: a45d9322-0d45-5dc9-b17b-32b843416c4b,
    },
    FunctionGUID {
        guid: a482559e-6a69-505e-b692-147deeced692,
    },
    FunctionGUID {
        guid: a603d813-2b33-567d-80c8-06e31562a400,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: a707a2d3-c7d2-5b93-97c9-1db77d8ccfe5,
    },
    FunctionGUID {
        guid: b9e4b6b0-47c3-54ca-9a01-56916ea0db8e,
    },
    FunctionGUID {
        guid: bae57993-863a-5c22-b8bb-92b862148af3,
    },
    FunctionGUID {
        guid: bfaf2f0a-d92d-5ef7-879d-86eb0ca01257,
    },
    FunctionGUID {
        guid: c3842e85-b9f2-54d5-b353-e9bb2a0b1203,
    },
    FunctionGUID {
        guid: c5766cb0-afbc-5701-ad2f-23d980f4bc15,
    },
    FunctionGUID {
        guid: cb8ebff6-30e8-5a8a-9c1b-ec5f26041d65,
    },
    FunctionGUID {
        guid: ccbd9943-6b24-59d8-ae4d-3ee868a2729c,
    },
    FunctionGUID {
        guid: ce28e080-0609-5bcd-abe1-2a6e57c91f83,
    },
    FunctionGUID {
        guid: cf4808fe-2fea-517d-8f85-f88ddc96cc78,
    },
    FunctionGUID {
        guid: e3b83b64-6ba1-5fe5-97da-5db987d79b07,
    },
    FunctionGUID {
        guid: eefbc8a9-f4f1-5725-9a4f-9dcd04ea3478,
    },
    FunctionGUID {
        guid: efcd394f-9201-57f7-aa2d-fc89ad355326,
    },
    FunctionGUID {
        guid: efcd394f-9201-57f7-aa2d-fc89ad355326,
    },
    FunctionGUID {
        guid: f25e7750-c61c-5ed8-a43d-959c3ba3a5b4,
    },
]
